kv/range_cache.go

```diff
@@ -137,82 +176,86 @@ func (rdc *rangeDescriptorCache) LookupRangeDescriptor(key roachpb.RKey,
 	} else if log.V(1) {
 		log.Infof("lookup range descriptor: key=%s", key)
 	}
-	rs, pErr := func(key roachpb.RKey, considerIntents, useReverseScan bool) ([]roachpb.RangeDescriptor, *roachpb.Error) {
-		// metadataKey is sent to rangeLookup to find the
-		// RangeDescriptor which contains key.
-		metadataKey, err := meta(key)
-		if err != nil {
-			return nil, roachpb.NewError(err)
+
+	var res lookupResult
+	requestKey := makeLookupRequestKey(key, considerIntents, useReverseScan)
+	rdc.lookupRequests.Lock()
+	if req, inflight := rdc.lookupRequests.inflight[requestKey]; inflight {
+		resC := make(chan lookupResult, 1)
+		req.observers = append(req.observers, resC)
+		rdc.lookupRequests.inflight[requestKey] = req
+		rdc.lookupRequests.Unlock()
+		rdc.rangeCacheMu.RUnlock()
+
+		res = <-resC
+	} else {
+		rdc.lookupRequests.inflight[requestKey] = req
+		rdc.lookupRequests.Unlock()
+		rdc.rangeCacheMu.RUnlock()
+
+		rs, pErr := rdc.performRangeLookup(key, considerIntents, useReverseScan)
+		if pErr != nil {
+			res = lookupResult{pErr: pErr}
+		} else if len(rs) == 0 {
+			res = lookupResult{pErr: roachpb.NewErrorf("no range descriptors returned for %s", key)}
+		} else {
+			res = lookupResult{desc: &rs[0]}
 		}
 
-		var (
-			// desc is the RangeDescriptor for the range which contains
-			// metadataKey.
-			desc *roachpb.RangeDescriptor
-			pErr *roachpb.Error
-		)
-
-		if bytes.Equal(metadataKey, roachpb.RKeyMin) {
-			// In this case, the requested key is stored in the cluster's first
-			// range. Return the first range, which is always gossiped and not
-			// queried from the datastore.
-			desc, pErr = rdc.db.FirstRange()
-			if pErr != nil {
-				return nil, pErr
+		rdc.rangeCacheMu.Lock()
+		if res.pErr == nil {
+			if err := rdc.insertRangeDescriptorsLocked(rs); err != nil {
+				res = lookupResult{pErr: roachpb.NewError(err)}
 			}
-			return []roachpb.RangeDescriptor{*desc}, nil
 		}
-		if bytes.HasPrefix(metadataKey, keys.Meta1Prefix) {
-			// In this case, desc is the cluster's first range.
-			if desc, pErr = rdc.db.FirstRange(); pErr != nil {
-				return nil, pErr
-			}
-		} else {
-			// Look up desc from the cache, which will recursively call into
-			// this function if it is not cached.
-			desc, pErr = rdc.LookupRangeDescriptor(metadataKey, considerIntents, useReverseScan)
-			if pErr != nil {
-				return nil, pErr
-			}
+
+		rdc.lookupRequests.Lock()
```

Comment:
Note that locking `rdc.lookupRequests` here isn't actually needed because we have an exclusive write lock on `rdc.rangeCacheMu`, but it makes things much easier to reason about. Should I keep it or just leave a comment?